docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+If it is set to zero, we *never* retry dialing a peer address.
+
+Upon each failed dial attempt, we increase the delay by `MinRetryTime`, plus an
+optional random jitter of up to `RetryTimeJitter`.
+
+The retry delay should not be longer than the `MaxRetryTime` parameter,
+or `MaxRetryTimePersistent` parameter in the case of persistent peers.
+
+#### Errors
+
+Errors are also returned if:
+
+- the updated peer information is invalid
+- there is an error when saving the peer state to the peer store
+
+### Accepted transition
+
+This transition is performed when the node establishes an *incoming* connection
+with a peer.
+This means that the router has received a connection attempt from this peer and
+successfully established a connection with it.
+
+It may occur, however, that when this transition is invoked the peer is already
+in the [`Connected`](#connected-peer) state.
+The most likely reason is that the router was simultaneously dialing the same
+peer, and has successfully [established a connection](#dialed-transition) with
+it.
+In this case, the transition fails, indicating to the router that is should
+close the accepted connection.
+
+It may also occur that the node is already connected to `MaxConnected` peers,
+which means that all connection slots are full.
+In this case, the peer manager tries to find a connection slot that can be
+[upgraded](#slots-upgrading) to give room for the accepted connection.
+If no suitable connection slot is found, or the hard limit `MaxConnected +
+MaxConnectedUpgrade` of connected peers is reached, the transitions fails.
+
+If the transition succeeds, the peer is set to the
+[`Connected`](#connected-peer) state as an `incoming` peer.
+
+The accepted peer might not be known by the peer manager.
+In this case the peer is registered in the peer store, without any associated
+address.
+The peer `LastConnected` time is set and the `DialFailures` counter is reset
+for all addresses associated to the peer.
+
+> If the peer is `Inactive`, it is set as active.
+> This action has not effect apart from producing metrics.
+
+If a connection slot was upgraded to give room for the accepted connection, the
+peer on that slot transitions to the [evict sub-state](#evict-peer).
+
+#### Errors
+
+The transition fails if:
+
+- the node accepted itself
+- the peer is already in the `Connected` state
+- the node is connected to `MaxConnected` peers, and no slot is suitable for upgrading
+
+Errors are also returned if:
+
+- the updated peer information is invalid
+- there is an error when saving the peer state to the peer store
+
+### Connected peer
+
+A peer to which the node is connected.
+A peer in this state is not considered a candidate peer.
+
+The peer manager distinguishes *incoming* from *outgoing* connections.
+Incoming connections are established through the [`Accepted`](#accepted-transition) transition.
+Outgoing connections are established through the [`Dialed`](#dialed-transition) transition.
+
```

Comment:
Another reflection.
If the peer manager takes the next candidate to dial (e.g. according to ranking) and it turns out it is an incoming peer, wouldn't it make sense to move it to the the set of outgoing peers without touching the connection?

This way, one extra incoming peer could be accepted later on